cartographer/mapping/sparse_pose_graph.cc

```diff
@@ -81,9 +107,23 @@ proto::SparsePoseGraphOptions CreateSparsePoseGraphOptions(
 proto::SparsePoseGraph SparsePoseGraph::ToProto() {
   proto::SparsePoseGraph proto;
 
+  const auto trajectory_nodes = GetTrajectoryNodes();
+  std::vector<const Submaps*> submap_pointers;
+  for (const auto& trajectory_node : trajectory_nodes) {
+    submap_pointers.push_back(trajectory_node.constant_data->trajectory);
+  }
+
+  const auto trajectory_indices = IndexTrajectories(submap_pointers);
+
   std::vector<std::vector<TrajectoryNode>> grouped_nodes;
-  std::vector<std::pair<int, int>> new_indices;
-  GroupTrajectoryNodes(GetTrajectoryNodes(), &grouped_nodes, &new_indices);
+  std::vector<std::pair<int, int>> grouped_node_indices;
+  GroupTrajectoryNodes(trajectory_nodes, trajectory_indices, &grouped_nodes,
+                       &grouped_node_indices);
+
+  std::vector<std::vector<SparsePoseGraph::SubmapState>> grouped_submaps;
+  std::vector<std::pair<int, int>> grouped_submap_indices;
+  GroupSubmapStates(GetSubmapStates(), trajectory_indices, &grouped_submaps,
```

Comment:
Did you consider improving this by making `GroupSubmapStates()` part of the interface instead of `GetSubmapStates()`? I think that would make sense. While it would mean that there are then three very similar implementations (the nodes variant, and a submap variant for each of 2D and 3D), it also removes quite a bit of code and would not expose internals as much.